pkg/controller/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -860,42 +869,60 @@ func (r *ReconcileAWSPrivateLink) reconcileVPCEndpoint(awsClient *awsClient,
 		Filters: []*ec2.Filter{tag},
 	}
 
-	var allVpcEndpoints []*ec2.VpcEndpoint
+	// HIVE-2837: Clean up redundant VPCEs (e.g. "Rejected")
+	var vpcEndpointsToDelete []*string
 	for {
 		resp, err := awsClient.hub.DescribeVpcEndpoints(input)
 		if err != nil {
 			endpointLog.WithError(err).Error("error getting VPC Endpoint")
-			return modified, nil, err
+			return endpointReconciled, nil, err
 		}
 
-		allVpcEndpoints = append(allVpcEndpoints, resp.VpcEndpoints...)
+		// HIVE-2838: First page may be empty even when there are results!
+		if len(resp.VpcEndpoints) != 0 {
+			for _, vpce := range resp.VpcEndpoints {
```

Comment:
Nit: Do we need the additional if wrapper? I thought `range` would simply skip when empty?

Reply:
Right you are. I feel like in earlier versions of go, an uninitialized list variable or struct member would throw a nil pointer panic under `range`. But I just verified with 1.23 and it's fine.